Avoid MultipleObjectsReturned errors with LoginFailures

efischer19 opened the pull request:
Initially reported and fixed by @laq in https://github.com/edx/edx-platform/pull/11313, replayed here after some git craziness. Original PR comments follow:

```
From this https://code.djangoproject.com/ticket/13906#no1 , the get_or_create function is vulnerable to race conditions in MySQL. This causes the LoginFailure to have, in some cases(rarely but happens), more than one row for the same user, thus breaking the login for the user with a MultipleObjectsReturned exception.

To avoid this I see two paths:

Either set the user field as unique, which I asume would make the get_or_create throw an exception when the race condition happens avoiding the duplicate row, but throwing an exception in the increment_lockout_counter method when that happens. This will require a migration to set the unique flag and a data migration to fix the current duplicate rows.

Or add an except for MultipleObjectsReturned on the is_user_locked_out method so when the duplicate row appears on the DB, it cleans the error, deleting the row or rows with oldest lockout dates. Leaving just one row for the user and solving the login problem.

Here I propose the second option, which would clean the duplicate rows and affect the rest of the code as little as possible.
```

edx-webhook commented:
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

efischer19 commented:
@laq, I had problems cherry-picking the previous commits after all the git craziness that happened overnight, so I've replayed them here to keep the ball rolling.

The last commit is the one I've proposed to:
- always perform this fix when `get()`-ing an object and
- expect to get multiple values occasionally, rather than using exceptions

Let me know what you think, and we'll get this squashed and merged post-haste.

laq commented:
Hi @efischer19, I looked at the change you suggest, and I like it.

efischer19 commented:
@laq Excellent! I've squashed all the commits and fixed a minor error (`records.first()` instead of `records[0]`, to handle empty QuerySets). Once tests have passed, we'll merge this to master.

Thanks again for the contribution, and sorry about all the git-dancing!

efischer19 commented:
Ack, returning `None` was also not the right move there. When this method is used, the surrounding code expects an object to be returned, or an `ObjectDoesNotExist` exception to be thrown. I've changed the last line of that helper to `records.get()`, and confirmed that locally passes the test that keeps failing.

efischer19 commented:
jenkins run bokchoy

laq commented:
@efischer19  Great!, Its nice to help. Git can be a pain some times.

efischer19 commented:
jenkins run bokchoy
